### ISSUE TITLE
Implement genotype confidence percentile

### ISSUE THREAD

**Issue by iqbal-lab** [rmcolq/pandora#216]
Need to do this to normalise gt-conf for overall covg of the sample.
This has been implemented in Mykrobe and minos (in python)

**Comment by iqbal-lab** [comment]
@martinghunt code is here
https://github.com/iqbal-lab-org/minos/blob/ed975d1099c6403eff79b04b0d2064eebfa95e73/minos/genotype_confidence_simulator.py

**Comment by iqbal-lab** [comment]
I did wonder about a single header and cpp file , which did this

**Comment by leoisl** [comment]
Am I right to assume this is for paper?

**Comment by iqbal-lab** [comment]
Happy to discuss tomorrow; is actually not too hard to implement. The decision when we designed the 4way, and when we had not implemented and tested GCP , was just to use same depth for all samples. Now, we have 24 samples at mildly varying different depths coming up, plus this issue in the 4way. So, the argument for doing this immediately is stronger.

**Comment by leoisl** [comment]
Do agree, it looks like the the cost is low for implementing this, with high benefits. However, other tools (e.g. snippy, nanopolish, medaka) I think do not implement such normalisation (they don't have the concept of multisample variant calling). If pandora does this normalisation, do we still need to present ROC curves on several subsampling depths (e.g. 30x, 60x and 100x)? Or do we just give the full data for each tool?

**Comment by iqbal-lab** [comment]
1. you don't need multisample var calling to need the normalisation - as soon as you use the same threshold for multiple single-sample VCFs, normalisation is an issue.
2. I don't think this matters. The plan for snippy/illumina is to give a baseline for comparison - how good an error rate you get with normal approaches, using our dataset and evaluation framework, and how high recall can go with one reference. 
3. Having GCP means we can define a meaningful threshold independent of coverage, so our samples have comparable filtering. It doesn't mean we perform the same independent of depth. We will still do better with more depth, certainly going up for 30-60x illumina anyway. *so the value of subsampling illumina is unchanged* - especially it means snippy performance will change. 
4. Having said that, i think there is diminishing returns, 90% of the value is obtained by just running at one depth, as mostly this section of the paper is to show how high recall can go with one ref.

**Comment by iqbal-lab** [comment]
Deleted previous rambling post

**Comment by iqbal-lab** [comment]
1. We definitely not doing 24 way illumina at multiple depths, too hard (many samples, different depths)
2. Subsampling and gcp are orthogonal
3. Snippy has essentially zero error, mad to worry about fairness of filtering
4. We can't normalise likelihoods of tools where we don't know how implemented. We ate fixing our own implementation limitation

**Comment by leoisl** [comment]
- [ ] Port https://github.com/iqbal-lab-org/minos/blob/ed975d1099c6403eff79b04b0d2064eebfa95e73/minos/genotype_confidence_simulator.py to C++ in a branch

**Comment by mbhall88** [comment]
I thought the idea was to have this as a separate "header-only" library? i.e. it shouldn't be a branch on `pandora`.  
I think we need to have another call to discuss the details of this implementation before we go galloping off in the wrong direction :horse_racing:

**Comment by leoisl** [comment]
Oh sorry. Not sure I understood all that you said previously, but I have an answer below (we indeed need to discuss this in more details very soon, as I am starting implementation now).

Header-only library is indeed less coupled and way more reusable, modularised and maintainable than some files in a branch. So it is indeed a better implementation design, but will take more time to do it in a proper way. For example, a core concept we will have in this header-only library is the Genotyper, which does not actually exist in pandora. In pandora, we have the concept of `VCF`s (class `VCF`) which has the ability to genotype itself. We would need to decouple these concepts by refactoring into at least two classes, `VCF` and `Genotyper`. We will also need to define a public interface to the `Genotyper` class that is not coupled with pandora concepts, so other users (e.g. `gramtools`) can implement this public interface and use it. Probably more issues will appear on the run...

In summary, I see three options:

1. **[Fastest but with bad design]** Just create a branch and do the genotype confidence percentile inside pandora itself (don't worry about reusability, don't change components interface, keep things tightly coupled, no refactoring). But we will have tests and make sure that genotype confidence percentile is being done correctly. This will get us from implementation to results very fast, can be convenient to finish the paper;
2. **[Medium ground]** Create a separate header-only library to deal with this, but don't do any refactoring. Basically, this header-only library will just be usable by pandora (e.g. no refactoring and things will be tightly coupled), but it will be a separated library;
3. **[Slowest but with good design]** Proper refactoring of pandora and creation of header-only library with proper specification of a `Genotyper` abstract class so that is usable to anyone who can concretely implement such `Genotyper` class;

Please add other options or correct any of these options. Our ultimate goal is always 3. I just don't know if we start by 1 for the sake of the paper, and then refactor the solution into 3 later. I also don't know how fast is 1 and how slow is 3.

**Comment by bricoletc** [comment]
I will be probably need GCP soon. So I'm biased towards option3. I don't particularly mind though because I know you're on the paper deadline ;-O

**Comment by iqbal-lab** [comment]
Yes, i'm reluctant to do option 3 before the paper, because a refactor could take a while, and generate more issues on the way. Let me expand on those options

0. Write a python script that normalises pandora VCFs. The simulation of SNPs could be the same as the minos code, but the genotyping would be basically whatever pandora does.

1. Leandro's branch implementation.

2.header file, but no refactoring. What does this actually mean? Would the following work as a solution?

simulate_snps(number of snps, genomewide average coverage, error rate, model)
// model can be neg binomial or poisson
returns a vector of pairs like this (covg on true allele, covg on wrong allele)

get_confidence_distribution(genomewide average coverage, error rate, vector of pairs)
 - calls the internal pandora genotyping code, gets the confidences, makes a list of confidences

make_confidence_to_percentile_map(list of confidences)
 - returns a map.

The first and third functions could exist in a separate library and would be reusable by gramtools.
The second function is coupled to pandora, but in a future update we could refactor and create a new library function that takes a genotyper class

3. complete refactor, glory of conan the barbarian rebuild civilisation.
do the above, with refactoring.

**Comment by mbhall88** [comment]
>     2. **[Medium ground]** Create a separate header-only library to deal with this, but don't do any refactoring. Basically, this header-only library will just be usable by pandora (e.g. no refactoring and things will be tightly coupled), but it will be a separated library;
> 
>     3. **[Slowest but with good design]** Proper refactoring of pandora and creation of header-only library with proper specification of a `Genotyper` abstract class so that is usable to anyone who can concretely implement such `Genotyper` class;
> 

I think both of these options are actually the same. What I mean by that is in both cases we implement a header-only GCP library. The only thing that differs is how `pandora` interacts with this library. And that is a separate issue altogether.  

It would be easy enough to patch an interface within `pandora` that allows us to pass a genotyper to the GCP library no?

![gcpint](https://user-images.githubusercontent.com/20403931/79343585-749aac80-7f26-11ea-9380-75d4222b5a06.png)

In this example dependency graph, the direction of the arrow indicates "depends on". Regardless of whether we go with Option 2 or 3 that you mentioned @leoisl , it would be best for us to implement an interface in this way so that there is no direct dependency between pandora and gcplib.

**Comment by leoisl** [comment]
> * Write a python script that normalises pandora VCFs. The simulation of SNPs could be the same as the minos code, but the genotyping would be basically whatever pandora does.
> * Leandro's branch implementation.

So these options are conflicting. The first is porting pandora genotyping code to python, and use what we have in minos. Second is porting minos GCP code to c++ and use pandora genotyping code. It is easier and less error prone to do the second.

> 2.header file, but no refactoring. What does this actually mean? Would the following work as a solution?

I vote to go for this option, that you explain here, with the change that the second function (i.e. the interaction between the GCP library and its user) is defined in an interface as explained by @mbhall88  in https://github.com/rmcolq/pandora/issues/216#issuecomment-614045916 . However, I won't worry much about this interface right now, we can refactor it as we need, and do the things in demand.

I think things are very abstract now, and we can easily get lost in the discussion. Will start a repo, and start something so that we can have something more concrete to discuss on.

**Comment by mbhall88** [comment]
I disagree. I think discussing these things beforehand is paramount. Otherwise, we get attached to code that we have written and say things like "let's just use what we've written so far for now and we'll change it later". 
If everyone else disagrees then I'm happy to back down.

**Comment by iqbal-lab** [comment]
I don't disagree. But let's try and clarify what we can do - i don't think we're that far from being able to do what you want.
I think this whole thing reduces to less than 5 functions. i listed 3 key ones above and suggested 2 would fit in this library and one be implemented by the caller. We could expand on your diagram Michael, and add in APIs and specify callers and responsibilities. This would give brice a chance to look at that and work out how it would/not work with gramtools (similar to minos), and leandro to do the same for pandora. How about that?

**Comment by iqbal-lab** [comment]
Basically i think this lib is simple enough that we can go right ahead and map out the full API, and then think. Something to simulate a set of SNPs (covg on two alleles), somethign to genotype those and get confidences, something to turn that into percentiles, something to get a gt-conf and return a percentile.
If you have those the caller can decide whether to apply to a VCF post hoc, or do it line by line, or do it before the VCF when you first genotype a site - the caller has the freedom to apply as they choose

**Comment by leoisl** [comment]
Yeah, you are right, we should design the full API before touching any line of code, I think I was just too eager to code already. Anyway, I think the best is to think that whatever code is developed before we have the proper API design is prototype and has the risk to be totally discarded (and we should do it). I have a first version of the `Genotyper` interface for you to take a look (making a PR at  https://github.com/leoisl/GCP). I think it is important to keep this interface as loosely coupled as possible, so it can be used with whatever caller.
The other concepts (Model, Simulator, etc) are simpler to design I think, as they are just going to be used by users (not extended and implemented and the `Genotyper` interface).
